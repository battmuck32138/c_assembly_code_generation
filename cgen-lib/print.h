// Used to print out strings.
int PrintString (char *str);

// Used to print out integers.
int PrintInt (int i);

// Used to print out character.
int PrintChar (char c);

// Prints bools as an integer
// value. Used solely for type
// matching.
int PrintBool (bool b);

// Prints out a newline
int PrintNewline ();

// Prints out many integers
int PrintInts (int a, int b, int c);

// Prints out many chars
int PrintChars (char a, char b, char c);

// Prints out chars and ints
int PrintMixed (char a, char b, int x, int y, char c, int z);
