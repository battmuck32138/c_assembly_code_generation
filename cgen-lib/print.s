PrintString:
  addi sp sp -4
  sw ra 0(sp)
  lw a1 4(sp) 
  addi a0 x0 4
  ecall
  jal PrintNewline
  lw ra 0(sp)
  addi sp sp 4
  jr ra

PrintInt:
  addi sp sp -4
  sw ra 0(sp)
  lw a1 4(sp) 
  addi a0 x0 1
  ecall
  jal PrintNewline
  lw ra 0(sp)
  addi sp sp 4
  jr ra

PrintChar:
  addi sp sp -4
  sw ra 0(sp)
  lb a1 4(sp)
  andi a1 a1 0xff 
  addi a0 x0 11
  ecall
  jal PrintNewline
  lw ra 0(sp)
  addi sp sp 4
  jr ra

PrintBool:
  addi sp sp -4
  sw ra 0(sp)
  lw a1 4(sp) 
  addi a0 x0 1
  ecall
  jal PrintNewline
  lw ra 0(sp)
  addi sp sp 4
  jr ra

PrintNewline:
  addi a1 x0 10 #Load newline char
  addi a0 x0 11
  ecall
  mv a0 x0
  jr ra

PrintInts:
  addi sp sp -8
  sw ra 4(sp)
  lw t0 8(sp)
  sw t0 0(sp)
  jal PrintInt
  lw t0 12(sp)
  sw t0 0(sp)
  jal PrintInt
  lw t0 16(sp)
  sw t0 0(sp)
  jal PrintInt
  lw ra 4(sp)
  addi sp sp 8
  jr ra

PrintChars:
  addi sp sp -8
  sw ra 4(sp)
  lb t0 8(sp)
  sb t0 0(sp)
  jal PrintChar
  lb t0 9(sp)
  sb t0 0(sp)
  jal PrintChar
  lb t0 10(sp)
  sb t0 0(sp)
  jal PrintChar
  lw ra 4(sp)
  addi sp sp 8
  jr ra
  

PrintMixed:
  addi sp sp -8
  sw ra 4(sp)
  lb t0 8(sp)
  sb t0 0(sp)
  jal PrintChar
  lb t0 9(sp)
  sb t0 0(sp)
  jal PrintChar
  lw t0 12(sp)
  sw t0 0(sp)
  jal PrintInt
  lw t0 16(sp)
  sw t0 0(sp)
  jal PrintInt
  lb t0 20(sp)
  sb t0 0(sp)
  jal PrintChar
  lw t0 24(sp)
  sw t0 0(sp)
  jal PrintInt
  lw ra 4(sp)
  addi sp sp 8
  jr ra
  
