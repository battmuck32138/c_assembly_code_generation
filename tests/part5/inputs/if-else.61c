#include "../../../cgen-lib/print.h"

int main () {
  if (true) {
    PrintString ("Success 1");
  }
  if (true) {
    PrintString ("Success 2");
  } else {
    PrintString ("Failure 2");
  }
  if (false) {
    PrintString ("Failure 3");
  }
  if (false) {
    PrintString ("Failure 4");
  } else {
    PrintString ("Success 4");
  }
  return 0;
}
