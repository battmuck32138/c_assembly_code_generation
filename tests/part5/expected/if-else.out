Success 1
Success 2
Success 4

Exited with error code 0
